generator/migration.go

```diff
@@ -0,0 +1,670 @@
+// the user to write a proper migration.
+type ManualChange struct {
+	Msg string
+}
+
+func (c *ManualChange) MarshalText() ([]byte, error) {
+	return []byte(fmt.Sprintf("+++ THIS REQUIRES MANUAL MIGRATION: %s +++\n", c.Msg)), nil
+}
+
+// SchemaDiff generates a change set with the diff between two schemas.
+func SchemaDiff(old, new *ModelSchema) ChangeSet {
+	var cs ChangeSet
+	for _, oldTable := range old.Tables {
+		if t := new.Table(oldTable.Name); t == nil {
+			cs = append(cs, &DropTable{Name: oldTable.Name})
+		} else {
+			cs = append(cs, TableSchemaDiff(oldTable, t)...)
+		}
+	}
+
+	for _, newTable := range new.Tables {
+		if t := old.Table(newTable.Name); t == nil {
+			cs = append(cs, &CreateTable{newTable})
+		}
+	}
+
+	return cs
+}
+
+// TableSchemaDiff generates a change set with the diff between two table
+// schemas.
+func TableSchemaDiff(old, new *TableSchema) ChangeSet {
+	var cs ChangeSet
+	for _, oldCol := range old.Columns {
+		if c := new.Column(oldCol.Name); c == nil {
+			cs = append(cs, &DropColumn{
+				Table: old.Name,
+				Name:  oldCol.Name,
+			})
+		} else {
+			cs = append(cs, ColumnSchemaDiff(old.Name, oldCol, c)...)
+		}
+	}
+
+	for _, newCol := range new.Columns {
+		if c := old.Column(newCol.Name); c == nil {
+			cs = append(cs, &AddColumn{
+				Table:  new.Name,
+				Column: newCol,
+			})
+		}
+	}
+	return cs
+}
+
+// ColumnSchemaDiff generates the change set with the diff between two column
+// schemas.
+func ColumnSchemaDiff(table string, old, new *ColumnSchema) ChangeSet {
+	var cs ChangeSet
+	if old.Type != new.Type {
+		cs = append(cs, &ManualChange{
+			fmt.Sprintf("don't know how to generate migration for a change of type in %s(%s)", table, new.Name),
+		})
+	}
+
+	if old.PrimaryKey != new.PrimaryKey {
+		cs = append(cs, &ManualChange{
+			fmt.Sprintf("don't know how to generate migration for a change of primary key in %s(%s)", table, new.Name),
+		})
+	}
+
+	if old.NotNull != new.NotNull {
+		cs = append(cs, &ManualChange{
+			fmt.Sprintf("don't know how to generate migration for a change of null/not null in %s(%s)", table, new.Name),
+		})
+	}
+
+	if referenceChanged(old, new) {
+		cs = append(cs, &ManualChange{
+			fmt.Sprintf("don't know how to generate migration for a change of foreign key in %s(%s)", table, new.Name),
+		})
+	}
+
+	return cs
+}
+
+func referenceChanged(old, new *ColumnSchema) bool {
+	return old.Reference != new.Reference &&
+		(old.Reference == nil ||
+			new.Reference == nil ||
+			old.Reference.Column != new.Reference.Column ||
+			old.Reference.Table != new.Reference.Table)
+}
+
+// ReverseChangeSet returns a new change set with the inverses of the given
+// change set.
+func ReverseChangeSet(cs ChangeSet, old *ModelSchema) ChangeSet {
+	var result = make(ChangeSet, len(cs))
+	for i, c := range cs {
+		result[i] = reverseChange(c, old)
+	}
+	return result
+}
+
+// reverseChange generates the inverse of a change.
+func reverseChange(c Change, old *ModelSchema) Change {
```

Comment:
Instead of do this, could be better create a method called `Reverse()` into the `Change` interface. Then each `Change` implementation can implement his inverse. Also, doing this, we can delete this `panic("unreachable")` that I hate so much.